rest_server/IOMgrConnection.go

```diff
@@ -0,0 +1,432 @@
+			} else if IOMgrChunkSize >= IOMgrChunkMaxBufferedSize {
+				IOMgrbuffSize = IOMgrChunkMaxBufferedSize
+				IOMgrChunkSize = IOMgrChunkSize - IOMgrChunkMaxBufferedSize
+			}
+			siodbLoggerPool.Debug("IOMgrbuffSize: %v", IOMgrbuffSize)
+
+			IOMgrbuff := make([]byte, uint32(len(IOMgrbuffRemain))+IOMgrbuffSize)
+
+			// Form JSONChunk
+			IOMgrBytesRead, err = io.ReadFull(IOMgrConn, IOMgrbuff[len(IOMgrbuffRemain):])
+			siodbLoggerPool.Debug("Bytes read from IOMgr: %v", IOMgrBytesRead)
+			if len(IOMgrbuffRemain) > 0 {
+				bc := copy(IOMgrbuff, IOMgrbuffRemain)
+				siodbLoggerPool.Debug("Bytes copied from IOMgrbuffRemain into IOMgrbuff : %v", bc)
+				IOMgrbuffRemain = nil
+			}
+
+			for len(IOMgrbuff) > 0 {
+				UTF8rune, size := utf8.DecodeRune(IOMgrbuff)
+				if UTF8rune == utf8.RuneError { // Truncated UTF8 read from IOMgr: keeps for next loop
+					IOMgrbuffRemain = make([]byte, len(IOMgrbuff))
+					IOMgrbuffRemain = IOMgrbuff
+					break
+				} else {
+					JSONChunk = JSONChunk + fmt.Sprintf("%c", UTF8rune)
+					IOMgrbuff = IOMgrbuff[size:]
+				}
+			}
+
+			// Bufferizing JSONChunk
+			JSONChunkBuff = JSONChunkBuff + JSONChunk
+			siodbLoggerPool.Debug("len(IOMgrbuffRemain): %v", len(IOMgrbuffRemain))
+			siodbLoggerPool.Debug("len(JSONChunk): %v", len(JSONChunk))
+			siodbLoggerPool.Debug("len(JSONChunkBuff): %v", len(JSONChunkBuff))
+			JSONChunk = ""
+
+			siodbLoggerPool.Debug("int(restServerConfig.HTTPChunkSize): %v", int(restServerConfig.HTTPChunkSize))
+			siodbLoggerPool.Trace("JSONChunkBuff: %s", JSONChunkBuff)
+			if len(JSONChunkBuff) >= int(restServerConfig.HTTPChunkSize) {
+				siodbLoggerPool.Debug("# len(JSONChunkBuff) > restServerConfig.HTTPChunkSize: sending to client...")
+				pos := int(0)
+				for i := 0; i <= (int(len(JSONChunkBuff)) - int(uint64(len(JSONChunkBuff))%restServerConfig.HTTPChunkSize) - int(restServerConfig.HTTPChunkSize)); i = i + int(restServerConfig.HTTPChunkSize) {
+					siodbLoggerPool.Debug("Steaming buffer from position %v to %v.", i, i+int(restServerConfig.HTTPChunkSize))
+					siodbLoggerPool.Trace("JSONChunkBuff Steamed: %s", JSONChunkBuff[i:i+int(restServerConfig.HTTPChunkSize)])
+					c.String(http.StatusOK, JSONChunkBuff[i:i+int(restServerConfig.HTTPChunkSize)])
+					pos = i
+				}
+				siodbLoggerPool.Debug("pos: %v", pos)
+				JSONChunkBuff = JSONChunkBuff[pos+int(restServerConfig.HTTPChunkSize):]
+				siodbLoggerPool.Debug("JSONChunkBuff leftovers size: %v", len(JSONChunkBuff))
+				siodbLoggerPool.Trace("JSONChunkBuff final leftovers: %s", JSONChunkBuff)
+			}
+
+		}
+
+	}
+
+	siodbLoggerPool.Debug("JSONChunkBuff final leftovers size: %v", len(JSONChunkBuff))
+	siodbLoggerPool.Trace("JSONChunkBuff final leftovers: %v", JSONChunkBuff)
+	c.String(http.StatusOK, JSONChunkBuff)
+
+	return nil
+
+}
+
+func (IOMgrConn *IOMgrConnection) readVarint32() (bytesRead int, ruint32 uint32, err error) {
+	var ruint64 uint64
+	if bytesRead, ruint64, err = IOMgrConn.readVarint(); err != nil {
```

Comment:
VERY BAD. Actually you can read varint64 and expose it as varint32.
Add parameter "maxInputBytes" to readVarint, pass 10 from readVarint64, pass 5 from readVarint32
In readVarint read up to maxInputBytes. If after this it has got 0x80 bit set in the maxInputBytes-th byte, this is error.